src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -622,80 +631,125 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
         {
-            await _writerLock.WaitAsync().ConfigureAwait(false);
-
-            // If the connection has been aborted, then fail now instead of trying to send more data.
-            if (IsAborted())
+            if (acquireWriteLock)
             {
-                throw new IOException(SR.net_http_invalid_response);
+                await AcquireWriteLockAsync(cancellationToken).ConfigureAwait(false);
             }
-        }
 
-        private void ReleaseWriteLock()
-        {
-            // Currently, we always flush the write buffer before releasing the lock.
-            // If we change this in the future, we will need to revisit this assert.
-            Debug.Assert(_outgoingBuffer.ActiveMemory.IsEmpty);
+            try
+            {
+                // If there is a pending write that was canceled while in progress, wait for it to complete.
+                if (_inProgressWrite != null)
+                {
+                    await _inProgressWrite;
+                    _inProgressWrite = null;
+                }
+
+                int totalBufferLength = _outgoingBuffer.TotalLength;
+                int activeBufferLength = _outgoingBuffer.ActiveSpan.Length;
 
-            _writerLock.Release();
+                if (totalBufferLength >= 32768 &&
+                    writeBytes >= totalBufferLength - activeBufferLength &&
+                    activeBufferLength > 0)
+                {
+                    // If the buffer has already grown to 32k, does not have room for the next request,
+                    // and is non-empty, flush the current contents to the wire.
+                    await FlushOutgoingBytesAsync().ConfigureAwait(false);
+                }
+                _outgoingBuffer.EnsureAvailableSpace(writeBytes);
+            }
+            catch
+            {
+                if (acquireWriteLock)
+                {
+                    _writerLock.Release();
+                }
+                throw;
+            }
         }
 
-        private async ValueTask SendSettingsAckAsync()
+        // This method handles flushing bytes to the wire. Writes here need to be atomic, so as to avoid
+        // killing the whole connection. Callers must hold the write lock, but can specify whether or not
+        // they want to release it.
+        private async ValueTask FinishWriteAsync(bool mustFlush, CancellationToken cancellationToken = default, bool releaseWriteLock = true)
         {
-            await AcquireWriteLockAsync().ConfigureAwait(false);
             try
             {
-                _outgoingBuffer.EnsureAvailableSpace(FrameHeader.Size);
-                WriteFrameHeader(new FrameHeader(0, FrameType.Settings, FrameFlags.Ack, 0));
+                // We must flush if the caller requires it, or if there are no other pending writes.
+                if (mustFlush || (releaseWriteLock && _pendingWriters == 0))
```

Comment:
What happens if there are pending writers but we read _pendingWriters as 0?  That appears to be possible given the synchronization involved; presumably if that happens it just means we flush when we wouldn't otherwise need to, and so worst case it's just a perf impact but not a functional impact, correct?

Reply:
That's correct. It's acceptable (though not ideal) to flush unnecessarily. The goal here was to err on the side of flushing, since failing to flush when there are no pending writers could cause us to never actually write the bytes to the wire.

If you see any potential ways to make this smarter about flushing though, don't hesitate to let me know.